lib/Dialect/FIRRTL/FIRRTLUtils.cpp

```diff
@@ -229,6 +226,199 @@ Value circt::firrtl::getModuleScopedDriver(Value val, bool lookThroughWires,
   return val;
 }
 
+bool circt::firrtl::walkDrivers(Value val, bool lookThroughWires,
+                                bool lookThroughNodes, bool lookThroughCasts,
+                                WalkDriverCallback callback) {
+  // TODO: what do we want to happen when there are flips in the type? Do we
+  // want to filter out fields which has reverse flow?
+  assert(val.getType().cast<FIRRTLType>().isPassive() &&
+         "this code was not tested with flips");
+
+  // This method keeps a stack of wires (or ports) and subfields of those that
+  // it still has to process.  It keeps track of which fields in the
+  // destination are attached to which fields of the source, as well which
+  // subfield of the source we are currently investigating.  The fieldID is
+  // used to filter which subfields of the current operation which we should
+  // visit. As an example, the src might be an aggregate wire, but the current
+  // value might be a subfield of that wire. The src field ref will represent
+  // all subaccesses to the target, but the fieldID for the current op only
+  // needs to represent the all subaccesses between the current op and the
+  // target.
+  struct StackElement {
+    StackElement(FieldRef dst, FieldRef src, Value current,
+                 Value::user_iterator it, Value::user_iterator end,
+                 unsigned fieldID)
+        : dst(dst), src(src), current(current), it(it), end(end),
+          fieldID(fieldID) {}
+    // The elements of the destination that this refers to.
+    FieldRef dst;
+    // The elements of the source that this refers to.
+    FieldRef src;
+
+    // These next fields are tied to the value we are currently iterating. This
+    // is used so we can check if a connect op is reading or driving from this
+    // value.
+    Value current;
+    // An iterator of the users of the current value.
+    Value::user_iterator it;
+    Value::user_iterator end;
+    // A filter for which fields of the current value we care about.
+    unsigned fieldID;
+  };
```

Comment:
I can probably reduce the number of things in a stack frame, e.g. for `dst` I think I only need the `dfst.fieldID` since the `dst.value` never changes, and `end` can be recreated on the fly from the `current` value. I could also replace the use of `current` and `fieldID` with a `FieldRef` but I personally found that confusing: usually a `FieldRef` points at a declaration, and specifically *not* a subfield operation.

Reply:
I made the change to stop storing the `end()` iterator in the struct, but the rest of the changes gave the code higher mental overhead, not sure if it's worth saving an extra pointer / stack element.